datafusion/expr/src/udf.rs

```diff
@@ -18,11 +18,30 @@
 //! Udf module contains foundational types that are used to represent UDFs in DataFusion.
 
 use crate::{Expr, ReturnTypeFunction, ScalarFunctionImplementation, Signature};
+use arrow::array::ArrayRef;
+use datafusion_common::Result;
 use std::fmt;
 use std::fmt::Debug;
 use std::fmt::Formatter;
 use std::sync::Arc;
 
+pub trait ScalarFunctionDef: Sync + Send + std::fmt::Debug {
```

Comment:
I really like using a trait to define functions. To be honest, I am not sure why DataFusion didn't so that originally. 

My only concern is that adding this we would now have *three* ways to define scalar functions (the ScalarUDF, `ScalarFunctionDef` and `BuiltInScalarFunction`). 

I wonder if this trait is needed, or can we extend `ScalarUDF` to account for aliases?

Or perhaps should we be aiming to consolidate *all* functions to use `ScalarFuntionDef` 🤔

Reply:
I think using a trait to define functions is more clear and easy to use, if extending `ScalarUDF`, then we must init a struct for each function in an imperative way, it can get messy if there are a lot of function packages to manage.
And `trait ScalarFunctionDef` should be equivalent to `ScalarUDF` and `BuitInScalarFunction`, it's possible to replace all with `ScalarFunctionDef` in the future
(I also updated a more detailed answer for this concern in the original PR rationale part)

Reply:
> I think using a trait to define functions is more clear and easy to use

Yes, I agree

> And trait ScalarFunctionDef should be equivalent to ScalarUDF 

Maybe we could do something like 

```rust
impl SclarFunctionDef for ScalarUDF {
...
}
```

And then (eventally) deprecate ScalarUDF as part of helping people migrate to using `ScalarFunctionUDF` 🤔